org-code-javabuilder/media/src/main/java/org/code/media/SoundLoader.java

```diff
@@ -10,6 +17,43 @@
    * @throws SoundException when the file cannot be found
    */
   public static double[] read(String filename) throws SoundException {
-    return null;
+    // Acquire AudioInputStream
+    final AudioInputStream audioInputStream = getAudioInputStream(filename);
+    final AudioFormat audioFormat = audioInputStream.getFormat();
+    if (!AudioUtils.isAudioFormatValid(audioFormat)) {
+      throw new SoundException("Invalid audio file format");
+    }
+
+    // Read audio file samples as bytes
+    final byte[] bytes;
+    try {
+      final int bytesToRead = audioInputStream.available();
+      bytes = new byte[bytesToRead];
+      final int bytesRead = audioInputStream.read(bytes);
+      if (bytesToRead != bytesRead) {
+        throw new SoundException("Error reading audio file");
+      }
+      audioInputStream.close();
+    } catch (IOException e) {
+      throw new SoundException(e);
+    }
+
+    return AudioUtils.convertByteArrayToDoubleArray(bytes, audioFormat.getChannels());
+  }
+
+  private static AudioInputStream getAudioInputStream(String filename) throws SoundException {
+    if (filename == null) {
+      throw new SoundException("Missing file name");
+    }
+
+    try {
+      final File file = new File(filename);
+      if (file.exists()) {
+        return AudioSystem.getAudioInputStream(file);
+      }
+      throw new SoundException("File does not exist: " + filename);
```

Comment:
I think we want to send keys (see Neighborhood's `ExceptionKey` class) so we can translate messages in the future, is that right @jmkulwik ?

Reply:
Yep good call - would we want to catch the `SoundException` in the theater package and add the key there, or do it here? It may get a little tricky with dependencies since theater depends on media here, so we wouldn't be able to use a Theater-specific key I think?

Reply:
I wonder if we should add a `MediaRuntimeException` like we do for neighborhood and theater? Curious for Jessie's thoughts since she did some work on our exception handling overall. Although I like the idea of having special exceptions in the method header for educational reasons.

Reply:
Agreed, we'll want to send keys - that's generally how we plan to translate and also how we plan to make more user-friendly error messages. I think it would make sense to add a Media Exception and keys of some sort. It should extend JavabuilderException or JavabuilderRuntimeException - depending on whether we want to require the student to handle it or not. It should be set up similar to NeighborhoodRuntimeException & ExceptionKeys from the neighborhood package (although the naming convention used for the ExceptionKeys class could probably use another rev).

Reply:
Gotcha - where does that leave `SoundException` then? Would we wrap this in a `MediaRuntimeException` with a key corresponding to audio and the message?

Reply:
Looking at this a bit more - I think I'll remove `SoundException` in favor of `MediaRuntimeException` and throw MediaRuntimeExceptions/InternalJavaBuilderExceptions accordingly.

Reply:
`SoundException` itself can inherit from `JavabuilderException`. Then you'd need to make `SoundExceptionKey` entries for each unique message that the user will see. i.e. `"Error reading audio file"` could be `INVALID_AUDIO_FILE` or something similar. The string itself would live on the client in the i18n folder for Java Lab: https://github.com/code-dot-org/code-dot-org/blob/staging/apps/i18n/javalab/en_us.json

Reply:
Ah okay cool, yeah that makes sense. Will add.

Reply:
`SoundException` is specified in some parts of the student-facing API, so we'll need to have it for at least those areas. I believe Theater is one of the places where students will be learning about how exceptions work, so the SoundException will facilitate that learning.

Reply:
Updated with exception keys for `SoundException`